src/memory/area_frame_allocator.rs

```diff
@@ -74,23 +80,72 @@ impl FrameAllocate for AreaFrameAllocator {
                 let address = area.base_addr + area.length - 1;
                 Frame::containing_address(address as usize)
             };
+            
+            let contained_by = |frame: &Frame, low: &Frame, high: &Frame| -> bool {
+               low <= frame && frame <= high 
+            };
 
-            if frame > current_area_last_frame {
-                // all frames of current area are used, switch to the new area
-                self.choose_next_area();
-            } else if frame >= self.kernel_start && frame <= self.kernel_end {
-                // 'frame' is used by the kernel
-                self.next_free_frame = Frame(self.kernel_end.0 + 1)
-            } else if frame >= self.multiboot_start && frame <= self.multiboot_end {
-                // 'frame' is used by the multiboot information structure
-                self.next_free_frame = Frame(self.multiboot_end.0 + 1)
-            } else {
-                // frame is unused, increment 'next_free_frame' and return it
+            let mut f_unused = true;
+            match frame {
+                ref f if &current_area_last_frame < f 
+                    // all frames of current area are used, switch to the new area
+                    => {
+                        self.choose_next_area();
+                        f_unused = false;
+                    },
+                ref f if contained_by(&frame, &self.kernel_start, &self.kernel_end) 
+                    // 'frame' is used by the kernel
+                    => {
+                        self.next_free_frame = Frame(self.kernel_end.0 + 1);
+                        f_unused = false;
+                    },
+                ref f if contained_by(&frame, &self.multiboot_start, &self.multiboot_end)
+                    // 'frame' is used by the multiboot information structure
+                    => {
+                        self.next_free_frame = Frame(self.multiboot_end.0 + 1);
+                        f_unused = false;
+                    },
+                ref f => {
+                    for module in self.boot_info.module_tags() {
+                        let start = module.start_address() as usize + KERNEL_BASE;
+                        let end = module.end_address() as usize + KERNEL_BASE;
+                        let startFrame = Frame::containing_address(start);
+                        let endFrame = Frame::containing_address(end);
+                        if contained_by(&frame, &startFrame, &endFrame) {
+                            // 'frame' is used by the multiboot structure
+                            self.next_free_frame = Frame(endFrame.0 + 1);
+                            f_unused = false;
+                        }
+                    }
+                },
+            };
+
+            if f_unused {
+                // hooray! we can allocate the frame!
                 self.next_free_frame.0 += 1;
                 return Some(frame);
+
+            } else {
```

Comment:
This addition is unnecessary. The `else` is not needed because the `if` returns.

Reply:
whoops! that seems to be a relic from an earlier version